crates/ruff_cli/src/commands/format.rs

```diff
@@ -155,47 +193,50 @@ fn format_path(
 
     // Format the source.
     match format_source(source_kind, source_type, Some(path), settings)? {
-        FormattedSource::Formatted(formatted) => {
-            if mode.is_write() {
+        FormattedSource::Formatted {
+            unformatted,
+            formatted,
+        } => match mode {
+            FormatMode::Write => {
                 let mut writer = File::create(path).map_err(|err| {
                     FormatCommandError::Write(Some(path.to_path_buf()), err.into())
                 })?;
                 formatted
                     .write(&mut writer)
                     .map_err(|err| FormatCommandError::Write(Some(path.to_path_buf()), err))?;
+                Ok(FormatResult::Formatted)
             }
-            Ok(FormatResult::Formatted)
-        }
+            FormatMode::Check => Ok(FormatResult::Formatted),
+            FormatMode::Diff => Ok(FormatResult::Diff {
+                path: path.to_path_buf(),
+                unformatted,
+                formatted,
+            }),
+        },
         FormattedSource::Unchanged(_) => Ok(FormatResult::Unchanged),
     }
 }
 
 #[derive(Debug)]
 pub(crate) enum FormattedSource {
     /// The source was formatted, and the [`SourceKind`] contains the transformed source code.
-    Formatted(SourceKind),
+    Formatted {
+        unformatted: SourceKind,
+        formatted: SourceKind,
+    },
```

Comment:
This is making me wonder if `format_source` should take `&SourceKind`, and we should change `Unchanged(SourceKind)` to just `Unchanged`, and remove `unformatted` from the `Formatted` variant. It strikes me as inflexible that the method takes ownership of `SourceKind` but only acts on references and then returns the owned `SourceKind` in all cases. But, this is minor -- if it doesn't work for whatever reason, or you disagree, it's fine.

Reply:
I guess the challenge with this is that we need to move `SourceKind` out of the linter, maybe into `ruff_source_file`? I recommend investigating this as a separate PR.

Reply:
Thought so too but didn't want to create so much churn, changed it now